payments/db/sql_store.go

```diff
@@ -342,6 +342,108 @@ func (s *SQLStore) loadRouteCustomRecords(ctx context.Context, db SQLQueries,
 	)
 }
 
+// paymentStatusBatchData holds lightweight resolution data for computing
+// payment status efficiently during deletion operations.
+type paymentStatusBatchData struct {
+	// resolutionTypes maps payment ID to a list of resolution types
+	// for that payment's HTLC attempts.
+	resolutionTypes map[int64][]sql.NullInt32
+}
+
+// loadPaymentResolutionsBatchData loads only HTLC resolution types for multiple
+// payments. This is a lightweight alternative to loadPaymentsBatchData that's
+// optimized for operations that only need to determine payment status.
+func (s *SQLStore) loadPaymentResolutionsBatchData(ctx context.Context,
+	db SQLQueries, paymentIDs []int64) (*paymentStatusBatchData, error) {
+
+	batchData := &paymentStatusBatchData{
+		resolutionTypes: make(map[int64][]sql.NullInt32),
+	}
+
+	if len(paymentIDs) == 0 {
+		return batchData, nil
+	}
+
+	// Fetch resolution types for all payments in a single batch query.
+	resolutions, err := db.FetchHtlcAttemptResolutionsForPayments(
+		ctx, paymentIDs,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch HTLC resolutions: %w",
+			err)
+	}
+
+	// Group resolutions by payment ID.
+	for _, res := range resolutions {
+		batchData.resolutionTypes[res.PaymentID] = append(
+			batchData.resolutionTypes[res.PaymentID],
+			res.ResolutionType,
+		)
+	}
+
+	return batchData, nil
+}
+
+// loadPaymentResolutions is a single-payment wrapper around
+// loadPaymentResolutionsBatchData for convenience and to prevent duplicate
+// queries.
+func (s *SQLStore) loadPaymentResolutions(ctx context.Context, db SQLQueries,
```

Comment:
Why would we have duplicates?

Reply:
was more referring to duplicate queries because when fetching the attempts for one payment no batching is needed, but we wrap it in the batching call.